Jchick/connector framework docs (#4429)
iTwin/itwinjs-core: docs/learning/WriteAConnector.md

    - [Locks & Codes](#locks--codes)
    - [More information](#more-information)

### Porting a connector

To port an existing Connector originally written on iModel.js 2.x, it will be necessary to:

> 1. Change the dependencies to the latest version.

Reviewer: latest version of what

Reviewer 2: > having a preview site of doc changes in the ci would be kinda neat @williamkbentley @ben-polinsky

I assume it can be done one way or another.

Reviewer 3: There is a CLI package available that allows you to build the doc site locally, if that would help. Other than that, the CI job does build the docs to validate links. It doesnt do anything with the output. I suppose we could put it somewhere.

Reviewer 2: Yeah we could just publish the output and download it since it is static or push it to netlify or something.

Author: @aruniverse just elaborated on which version of what.  Hope that makes better sense now.

> 2. Search and replace "iModelBridge" with "BaseConnector", "BridgeJobDefArgs" with "JobArgs" and "BridgeRunner" with "ConnectorRunner" and so on.

Author: Cleaned up the language to be more precise.

#### Briefcases

> [Briefcases](./glossary#briefcase) are the local copies of iModel that users can acquire to work with the iModel. A Connector will download a briefcase locally using the ConnectorRunner and change their copy of iModel. Once all the work is done, the results are then pushed into the iModel. Please see the section on [Execution sequence](#execution-sequence) on the different steps involved.

Reviewer 2: _I know this doesn't pertain to this PR, but might as well clean this up_
This is a bit confusing due to inconsistent pluralization.

```suggestion
[Briefcases](./glossary#briefcase) are the local copies of iModels users can acquire. A Connector will download a briefcase locally using the ConnectorRunner and change the local copy of their iModel. Once all the work is done, the results are then pushed into the iModel. Please see the section on [Execution sequence](#execution-sequence) on the different steps involved.
```

Author: @ben-polinsky thanks for pointing this out - I agree that it was confusing.  I tried to make it more efficient by simply saying briefcase is a local copy and then refer to the briefcase w/o saying local or imodel.  I think it is straightforward now.

### Getting started

You'll need Node.js version ">=14.17.0 <17.0”. Please refer to [Getting Started](../getting-started/index#getting-started) for more details.

The node packages you'll need can be installed using

```Shell
$npm install  @itwin/core-backend
$npm install  @itwin/core-common
$npm install  @itwin/core-bentley
$npm install  @itwin/node-cli-authorization

$npm install  --save-dev @itwin/build-tools

> $npm install  --save-dev @itwin/core-backend

Reviewer: so core bentley & backend are both a real + dev dep?

Reviewer: can someone review if the deps here make sense for ppl writing their own connector? like the testing, linting, pkgs, etc

Author: @aruniverse just updated the npm install to match an actual connector successfully using the new connector-framework.

This method is the main workhorse of your Connector. When UpdateExistingData is called, models and definitions should be created and available for insertion of elements (if that is the desired workflow). Note: In the example below, definition elements are being inserted into the definition model at this point as well. Physical elements and Group elements can now be converted.

```ts
[[include:TestConnector-updateExistingData.example-code]]
```

### Execution Sequence

> The ultimate purpose of a Connector is to synchronize an iModel with the data in one or more source documents. The synchronization step involves authorization, communicating with an iModel server, converting data, and concurrency control. iTwin.js defines a framework in which the Connector itself can focus on the tasks of extraction, alignment, and change-detection. The other functions are handled by classes provided by iTwin.js. The framework is implemented by the ConnectorRunner class. A ConnectorRunner conducts the overall synchronization process. It loads and calls functions on a Connector at the appropriate points in the sequence. The process may be summarized as follows:

Reviewer: "iModel server" not sure how i feel about this, but i understand this was what we used before too

Author: @aruniverse replaced w iModelHub after discussion w/ my teammates.
